tutorials/dataframe/df106_HiggsToFourLeptons.py

```diff
@@ -0,0 +1,194 @@
+## \notebook -draw
+## This tutorial is the Higgs to four lepton analysis from the ATLAS Open Data release in 2020
+## (http://opendata.atlas.cern/release/2020/documentation/). The data was taken with the ATLAS detector
+## during 2016 at a center-of-mass energy of 13 TeV. The decay of the Standard Model Higgs boson
+## to two Z bosons and subsequently to four leptons is called the "golden channel". The selection leads
+## to a narrow invariant mass peak on top a relatively smooth and small background, revealing
+## the Higgs at 125 GeV.
+##
+## The analysis is translated to a RDataFrame workflow processing about 300 MB of simulated events and data.
+##
+## \macro_image
+## \macro_code
+## \macro_output
+##
+## \date March 2020
+## \author Stefan Wunsch (KIT, CERN)
+
+import ROOT
+import json
+import os
+
+# Create a ROOT dataframe for each dataset
+# Note that we load the filenames from an external json file.
+path = "root://eospublic.cern.ch//eos/opendata/atlas/OutreachDatasets/2020-01-22"
+files = json.load(open(os.path.join(os.environ["ROOTSYS"], "tutorials/dataframe", "df106_HiggsToFourLeptons.json")))
+processes = files.keys()
+df = {}
+xsecs = {}
+sumws = {}
+samples = []
+for p in processes:
+    for d in files[p]:
+        # Construct the dataframes
+        folder = d[0] # Folder name
+        sample = d[1] # Sample name
+        xsecs[sample] = d[2] # Cross-section
+        sumws[sample] = d[3] # Sum of weights
+        samples.append(sample)
+        df[sample] = ROOT.RDataFrame("mini", "{}/4lep/{}/{}.4lep.root".format(path, folder, sample))
+
+# Select events for the analysis
+ROOT.gInterpreter.Declare("""
+using VecF_t = const ROOT::RVec<float>&;
+using VecI_t = const ROOT::RVec<int>&;
+bool GoodElectronOrMuon(VecI_t type, VecF_t pt, VecF_t eta, VecF_t phi, VecF_t e, VecF_t trackd0pv, VecF_t tracksigd0pv, VecF_t z0)
+{
+    for (size_t i = 0; i < type.size(); i++) {
+        ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+        if (type[i] == 11) {
+            if (pt[i] < 7000 || abs(eta[i]) > 2.47 || abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        } else {
+            if (abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        }
+    }
+    return true;
+}
+""")
+
+for s in samples:
+    # Select electron or muon trigger
+    df[s] = df[s].Filter("trigE || trigM")
+
+    # Select events with exactly four good leptons conserving charge and lepton numbers
+    df[s] = df[s].Define("good_lep", "(abs(lep_eta) < 2.5) && (lep_pt > 5000) && (lep_ptcone30 / lep_pt < 0.3) && (lep_etcone20 / lep_pt < 0.3)")\
+                 .Filter("Sum(good_lep) == 4")\
+                 .Filter("Sum(lep_charge[good_lep]) == 0")\
+                 .Define("goodlep_sumtypes", "Sum(lep_type[good_lep])")\
+                 .Filter("goodlep_sumtypes == 44 || goodlep_sumtypes == 52 || goodlep_sumtypes == 48")
```

Comment:
A little explanation maybe why summing over lep types results in those magic numbers?

Reply:
hm i never know how much physics knowledge you can assume. so where to start and where to stop with explanations regarding what exactly is done?

Reply:
After sleeping one night over this: I think the average physicist can add up 4 * 11, 4 * 13 and 2 * 11 + 2 * 13 ;)

Reply:
could you give filters a descriptive name? that might go a long way and does not require adding lengthy comments

Reply:
I've added a comment. I don't really like giving the filters a name if there's no report. It makes it look messy a little :/

Reply:
added a comment